File: leetcode/recursion/shortest_path_binary_matrix.rb
# https://leetcode.com/problems/shortest-path-in-binary-matrix/
#
#
#
# @param {Integer[][]} grid
# @return {Integer}
#
require 'byebug'
def shortest_path_binary_matrix(grid)
  m = grid.length - 1 || 0
  n = m && grid[0].length - 1

  return -1 if m == 0 || n == 0 || grid[0][0] == 1 || grid[m][n] == 1

  dirs =[[-1, -1], [-1, 0], [-1, 1], [0, 1], [1, 1], [1, 0], [1, -1], [0, -1]]
  queue =[[0,0,1]] #x,y,len

  while (queue.length > 0)
    x,y,len = queue.pop
    puts len
    return len if x == m && y == m
    dirs.each do |dir|
      dx, dy = dir[0],dir[1]
      row = x + dx
      column = y + dy

      puts "#grid[#{row}][#{column}] #{grid[row][column]}"

      # if (row >= 0 && column >= 0 && row < grid.length && column < grid[0].length && !grid[row][column])
      #   queue.push([row, column, len + 1])
      #   grid[row][column] = 1
      # end

      # if within_cell(row, column,  m, n) && grid[row][column] == 0
      #   queue.push([row, column, len+1])
      #   grid[row][column]=1
      # end
    end
  end
  return -1
end

def within_cell(x,y, m,n)
  0 <= x && 0 <= y && x < m && y < n
end

grid = [[0, 0, 0], [1, 1, 0], [1, 1, 0]]
puts shortest_path_binary_matrix(grid)

File: leetcode/searching/shortestPathBinaryMatrix.js
var shortestPathBinaryMatrix = function(grid) {
    let q = [[0,0,0]];

    while(q.length > 0) {
        let next = q.shift();
        console.log(next)
        let [i,j,len] = next;
        console.log('i ', i)
        console.log('j ', j)
        console.log('len ', len)
        if(i === grid.length && j === grid[0].length) {
            return len;
        }
        let val = grid[i] && grid[i][j];
        console.log('val', val)
        if(val === 0) {
            grid[i][j] = 1;
            q.push([i+1,j,len+1]);
            q.push([i,j+1,len+1]);
            q.push([i+1,j+1,len+1]);

            q.push([i-1,j,len+1]);
            q.push([i,j-1,len+1]);
            q.push([i-1,j-1,len+1]);

            q.push([i-1,j+1,len+1]);
            q.push([i+1,j-1,len+1]);
            console.log('q', q)
        }
    }

    return -1;
};

// var shortestPathBinaryMatrix = function(grid) {
//     if(grid[0][0] == 1) return -1;
//     if(grid[grid.length - 1][grid[0].length - 1] == 1) return -1;
//
//     let queue = [];
//     queue.push([0, 0, 1]); // x, y, depth
//     while (queue.length > 0){
//         let node = queue.shift();
//         if(node[0] == grid.length - 1 && node[1] == grid.length - 1) return node[2];
//         for (let i = -1; i < 2; i++){
//             for (let j = -1; j < 2; j++){
//                 if(grid.length > node[0] + i && node[0] + i >= 0 && grid.length > node[1] + j && node[1] + j >= 0){
//                     if(grid[node[0] + i][node[1] + j] == 0){
//                         grid[node[0] + i][node[1] + j] = 1;
//                         queue.push([node[0] + i, node[1] + j, node[2] + 1]);
//                     }
//                 }
//             }
//         }
//     }
//     return -1;
// };
//
// console.log(shortestPathBinaryMatrix([[0, 0, 0], [1, 1, 0], [1, 1, 0]]))
console.log(shortestPathBinaryMatrix([[0,0,0],[1,1,0],[1,1,0]]))